### PR TITLE
[WIP] Build nixpkgs with gcc 8 by default

### PR DESCRIPTION
**Pull request by fpletz**
This PR bumps gcc used in `stdenv` to version 8. The main motivation for this is the ability to add `-fstack-clash-protection` to our default hardening flags (#53753). This will be done in a separate PR.

I anticipate quite a few of broken packages because there are some new checks, i.e. for array bounds and format truncation.

There is also an old `gcc8` branch which I noticed just now thanks to @Synthetica9. I'll cherry-pick changes from there.

Hydra Job: https://hydra.nixos.org/jobset/nixpkgs/gcc8
Newly failing builds with gcc8: https://hydra.nixos.org/eval/1500732?compare=staging-next

**Current state:** Most packages seem to build. `clang` segfaults currently. `gcc` fails on Darwin.

### CODE REVIEW

File: pkgs/tools/networking/dhcp/default.nix [hunk 1]
```diff
@@ -33,7 +33,7 @@ stdenv.mkDerivation rec {
     "--localstatedir=/var"
   ] ++ stdenv.lib.optionals (openldap != null) [ "--with-ldap" "--with-ldapcrypto" ];
 
-  NIX_CFLAGS_COMPILE = [ "-Wno-error=pointer-compare" ];
+  NIX_CFLAGS_COMPILE = [ "-Wno-error=pointer-compare" "-Wno-error=format-truncation" ];
```
Comment (Mic92) [inline pkgs/tools/networking/dhcp/default.nix]: Howe about `-Wno-error`?

File: pkgs/development/libraries/libomxil-bellagio/default.nix [hunk 2]
```diff
@@ -14,6 +14,8 @@ stdenv.mkDerivation rec {
 
   patches = [ ./fedora-fixes.patch ];
 
+  NIX_CFLAGS_COMPILE = [ "-Wno-error=array-bounds" ];
```
Comment (Mic92) [inline pkgs/development/libraries/libomxil-bellagio/default.nix]: here to.

File: pkgs/tools/networking/dhcp/default.nix (hunk 1 above)
Comment (globin) [reply pkgs/tools/networking/dhcp/default.nix]: I'd prefer this explicit error disabling.

**Comment by edolstra** [comment]
I see 172 commits, maybe you need to rebase this?

**Comment by vcunat** [comment]
I believe it's mainly because it got master commits that are missing from staging at this point.

**Comment by FRidh** [comment]
master and staging-next have been merged into staging so it can be rebased

File: pkgs/os-specific/linux/libselinux/default.nix [hunk 3]
```diff
@@ -35,7 +35,6 @@ stdenv.mkDerivation rec {
     "MAN3DIR=$(man)/share/man/man3"
     "MAN5DIR=$(man)/share/man/man5"
     "MAN8DIR=$(man)/share/man/man8"
-    "PYSITEDIR=$(py)/${python.sitePackages}"
```
Comment (jtojnar) [inline pkgs/os-specific/linux/libselinux/default.nix]: You can use `PYTHONLIBDIR`: https://github.com/SELinuxProject/selinux/blob/dc3d9c6d7173354cc4caf225ce75c41239cf47bd/libselinux/src/Makefile#L17

Comment (fpletz) [reply pkgs/os-specific/linux/libselinux/default.nix]: Oh, yeah, thanks! Will fix this in the next push.

**Comment by matthewbauer** [comment]
Let's aim for this right after the 19.03 branch off. We still have quite a few things in staging that I am wanting to prioritize.

**Comment by matthewbauer** [comment]
Sorry didn't mean to close it!

**Comment by vcunat** [comment]
Why the hell was f959c99 detected as merged to staging?  It certainly doesn't seem reachable ATM.

**Comment by vcunat** [comment]
Well, GitHub still thinks it's merged and thus won't let me reopen in-place.  Instead I opened https://github.com/NixOS/nixpkgs/pull/63132 so we don't lose the work.

**Pull request merged**